# Zomato Data Cleaning
# You have to perform the following tasks:
##  Data Cleaning:

- **Deleting redundant columns.**
- **Renaming the columns.**
- **Dropping duplicates.**
- **Cleaning individual columns.**
- **Remove the NaN values from the dataset**
- **Check for some more Transformations**

In [1]:
# importing the required libraries!

import pandas as pd
import numpy as np

In [2]:
# reading the zomato data, which are in the csv format!

z_data = pd.read_csv('zomato.csv')

In [9]:
#look some data from the top
z_data.head()

url   
0  https://www.zomato.com/bangalore/jalsa-banasha...  \
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone   
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233  \
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type   
0  Banashankari        Casual Dining  \
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked   
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...  \
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)   
0  North Indian, Mughlai, Chinese                         800  \
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item   
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []  \
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [8]:
# look some data from the from the bottom
z_data.tail()

url   
51712  https://www.zomato.com/bangalore/best-brews-fo...  \
51713  https://www.zomato.com/bangalore/vinod-bar-and...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   
51715  https://www.zomato.com/bangalore/chime-sherato...   
51716  https://www.zomato.com/bangalore/the-nest-the-...   

                                                 address   
51712  Four Points by Sheraton Bengaluru, 43/3, White...  \
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name online_order   
51712  Best Brews - Four Points by Sheraton Bengaluru...           No  \
51713                           Vinod Bar And Restaurant           No   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...           No   
51716                       The Nest - The Den Bengaluru           No   

      book_table    rate  votes           phone                    location   
51712         No  3.6 /5     27    080 40301477                  Whitefield  \
51713         No     NaN      0  +91 8197675843                  Whitefield   
51714         No     NaN      0             NaN                  Whitefield   
51715        Yes  4.3 /5    236    080 49652769  ITPL Main Road, Whitefield   
51716         No  3.4 /5     13  +91 8071117272  ITPL Main Road, Whitefield   

                rest_type                    dish_liked   
51712                 Bar                           NaN  \
51713                 Bar                           NaN   
51714                 Bar                           NaN   
51715                 Bar  Cocktails, Pizza, Buttermilk   
51716  Bar, Casual Dining                           NaN   

                                     cuisines approx_cost(for two people)   
51712                             Continental                       1,500  \
51713                             Finger Food                         600   
51714                             Finger Food                       2,000   
51715                             Finger Food                       2,500   
51716  Finger Food, North Indian, Continental                       1,500   

                                            reviews_list menu_item   
51712  [('Rated 5.0', "RATED\n  Food and service are ...        []  \
51713                                                 []        []   
51714                                                 []        []   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...        []   
51716  [('Rated 5.0', 'RATED\n  Great ambience , look...        []   

      listed_in(type) listed_in(city)  
51712   Pubs and bars      Whitefield  
51713   Pubs and bars      Whitefield  
51714   Pubs and bars      Whitefield  
51715   Pubs and bars      Whitefield  
51716   Pubs and bars      Whitefield

# Checking for Na's in an indivisual columns

In [140]:
order_z_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   name                         51717 non-null  object
 1   location                     51717 non-null  object
 2   address                      51717 non-null  object
 3   phone                        51717 non-null  object
 4   online_order                 51717 non-null  object
 5   cuisines                     51717 non-null  object
 6   listed_in(type)              51717 non-null  object
 7   rate                         43942 non-null  object
 8   votes                        51717 non-null  int64 
 9   book_table                   51717 non-null  object
 10  rest_type                    51490 non-null  object
 11  approx_cost(for two people)  51371 non-null  object
dtypes: int64(1), object(11)
memory usage: 4.7+ MB


In [18]:
# checking in menu_item column, how many null value in this particular column.

# checking the length of value in menu_item column
# by seeing the length of the menu_item column I deicede to drop this column because here there is so many empty row.
length = z_data['menu_item'].str.len()

In [21]:
length.head()

0    2
1    2
2    2
3    2
4    2
Name: menu_item, dtype: int64

In [116]:
# checking in location column 

emp_location = order_z_data['location'].isnull().sum()
emp_location

21

In [120]:
# checking in address column
emp_address = order_z_data['address'].isnull().sum()
emp_address

0

In [122]:
# checking in online_order
emp_online_order = order_z_data['online_order'].isnull().sum()
emp_online_order

0

In [124]:
# checking in cuisines column
emp_cuisines = order_z_data['cuisines'].isnull().sum()
emp_cuisines

45

In [128]:
# checking in listed_in(type) column
emp_listed = order_z_data['listed_in(type)'].isnull().sum()
emp_listed


0

In [130]:
# checking for rate column
emp_rate = order_z_data['rate'].isnull().sum()
emp_rate

7775

In [132]:
# checking in votes column
emp_votes = order_z_data['votes'].isnull().sum()
emp_votes

0

In [133]:
# checking in book_table column
emp_book_t = order_z_data['book_table'].isnull().sum()
emp_book_t

0

In [134]:
# checking in rest_type column
emp_rest_type = order_z_data['rest_type'].isnull().sum()
emp_rest_type

227

In [135]:
# checking in approx_cost(for two people) column
emp_cost = order_z_data['approx_cost(for two people)'].isnull().sum()
emp_cost

346

## Droping columns of the dataframe 

- Here I observe that the menu_item and review_list is useless for now!

In [28]:
# droping some useless columns from the datafram for now
new_z_data = z_data.drop(['reviews_list','menu_item','url','listed_in(city)'],axis = 1) 
new_z_data.head()

address                   name   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone   
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233  \
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type   
0  Banashankari        Casual Dining  \
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked   
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...  \
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet  
1     Chinese, North Indian, Thai                         800          Buffet  
2          Cafe, Mexican, Italian                         800          Buffet  
3      South Indian, North Indian                         300          Buffet  
4        North Indian, Rajasthani                         600          Buffet

# Reorder the columns for better understanding

In [65]:
columns_reorder = ['name','location','address','phone','online_order','cuisines','listed_in(type)','rate','votes','book_table','rest_type','approx_cost(for two people)']
order_z_data = z_data[columns_reorder]
order_z_data.head()

name      location   
0                  Jalsa  Banashankari  \
1         Spice Elephant  Banashankari   
2        San Churro Cafe  Banashankari   
3  Addhuri Udupi Bhojana  Banashankari   
4          Grand Village  Basavanagudi   

                                             address   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2  1112, Next to KIMS Medical College, 17th Cross...   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   

                              phone online_order   
0    080 42297555\r\n+91 9743772233          Yes  \
1                      080 41714161          Yes   
2                    +91 9663487993          Yes   
3                    +91 9620009302           No   
4  +91 8026612447\r\n+91 9901210005           No   

                         cuisines listed_in(type)   rate  votes book_table   
0  North Indian, Mughlai, Chinese          Buffet  4.1/5    775        Yes  \
1     Chinese, North Indian, Thai          Buffet  4.1/5    787         No   
2          Cafe, Mexican, Italian          Buffet  3.8/5    918         No   
3      South Indian, North Indian          Buffet  3.7/5     88         No   
4        North Indian, Rajasthani          Buffet  3.8/5    166         No   

             rest_type approx_cost(for two people)  
0        Casual Dining                         800  
1        Casual Dining                         800  
2  Cafe, Casual Dining                         800  
3          Quick Bites                         300  
4        Casual Dining                         600

In [66]:
order_z_data.tail()

name   
51712  Best Brews - Four Points by Sheraton Bengaluru...  \
51713                           Vinod Bar And Restaurant   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...   
51716                       The Nest - The Den Bengaluru   

                         location   
51712                  Whitefield  \
51713                  Whitefield   
51714                  Whitefield   
51715  ITPL Main Road, Whitefield   
51716  ITPL Main Road, Whitefield   

                                                 address           phone   
51712  Four Points by Sheraton Bengaluru, 43/3, White...    080 40301477  \
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...  +91 8197675843   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...             NaN   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...    080 49652769   
51716  ITPL Main Road, KIADB Export Promotion Industr...  +91 8071117272   

      online_order                                cuisines listed_in(type)   
51712           No                             Continental   Pubs and bars  \
51713           No                             Finger Food   Pubs and bars   
51714           No                             Finger Food   Pubs and bars   
51715           No                             Finger Food   Pubs and bars   
51716           No  Finger Food, North Indian, Continental   Pubs and bars   

         rate  votes book_table           rest_type   
51712  3.6 /5     27         No                 Bar  \
51713     NaN      0         No                 Bar   
51714     NaN      0         No                 Bar   
51715  4.3 /5    236        Yes                 Bar   
51716  3.4 /5     13         No  Bar, Casual Dining   

      approx_cost(for two people)  
51712                       1,500  
51713                         600  
51714                       2,000  
51715                       2,500  
51716                       1,500

# Percentage of Na's value in our dataframe

In [67]:
num_cell = np.product(order_z_data.shape)
num_cell

620604

In [68]:
count_emp_cell = order_z_data.isnull().sum().sum()
count_emp_cell

9622

In [69]:
percent_emp_cell = round((count_emp_cell/num_cell)*100,4)
print(f"{percent_emp_cell} %")

1.5504 %


In [70]:
order_z_data['name'].count()

51717

In [72]:
sum_dish_liked =z_data['dish_liked'].isnull().sum()

In [73]:
cell_with_value =  z_data.shape[0] - sum_dish_liked
cell_with_value

23639

In [75]:
#checking for phone column

phone_emp = z_data['phone'].isnull().sum()
phone_emp

1208

# Filling the missing values with another values

#### As, I noticed here in phone column there around 1208 row are empty and this not an easy or you can say it is impossible task for now to get this amount of phone numbers
- So, Here I simply going to replace these values with 'not mentioned(NM)' for short

In [114]:
order_z_data['phone'].fillna(value = 'NM', inplace = True) # inplace is used to change the value at original dataset

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/3543354556.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['phone'].fillna(value = 'NM', inplace = True) # inplace is used to change the value at original dataset


In [113]:
order_z_data

name   
0                                                  Jalsa  \
1                                         Spice Elephant   
2                                        San Churro Cafe   
3                                  Addhuri Udupi Bhojana   
4                                          Grand Village   
...                                                  ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...   
51713                           Vinod Bar And Restaurant   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...   
51716                       The Nest - The Den Bengaluru   

                         location   
0                    Banashankari  \
1                    Banashankari   
2                    Banashankari   
3                    Banashankari   
4                    Basavanagudi   
...                           ...   
51712                  Whitefield   
51713                  Whitefield   
51714                  Whitefield   
51715  ITPL Main Road, Whitefield   
51716  ITPL Main Road, Whitefield   

                                                 address   
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                  phone online_order   
0        080 42297555\r\n+91 9743772233          Yes  \
1                          080 41714161          Yes   
2                        +91 9663487993          Yes   
3                        +91 9620009302           No   
4      +91 8026612447\r\n+91 9901210005           No   
...                                 ...          ...   
51712                      080 40301477           No   
51713                    +91 8197675843           No   
51714                                NM           No   
51715                      080 49652769           No   
51716                    +91 8071117272           No   

                                     cuisines listed_in(type)    rate  votes   
0              North Indian, Mughlai, Chinese          Buffet   4.1/5    775  \
1                 Chinese, North Indian, Thai          Buffet   4.1/5    787   
2                      Cafe, Mexican, Italian          Buffet   3.8/5    918   
3                  South Indian, North Indian          Buffet   3.7/5     88   
4                    North Indian, Rajasthani          Buffet   3.8/5    166   
...                                       ...             ...     ...    ...   
51712                             Continental   Pubs and bars  3.6 /5     27   
51713                             Finger Food   Pubs and bars     NaN      0   
51714                             Finger Food   Pubs and bars     NaN      0   
51715                             Finger Food   Pubs and bars  4.3 /5    236   
51716  Finger Food, North Indian, Continental   Pubs and bars  3.4 /5     13   

      book_table            rest_type approx_cost(for two people)  
0            Yes        Casual Dining                         800  
1             No        Casual Dining                         800  
2             No  Cafe, Casual Dining                         800  
3             No          Quick Bites                         300  
4             No        Casual Dining                         600  
...          ...                  ...                         ...  
51712         No                  Bar                  

In [121]:
#validating by again checking the phone columns
order_z_data['phone'].isnull().sum()

0

**Again I observed that in location column 21 rows are NaN value** 
- I decided to replace with next row value. as I observe in data most of the location are repeating with certain pattern
- I am bfill() method to do so.

In [117]:
order_z_data['location'].fillna(method = 'bfill', limit = 1, inplace = True)

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/2964109379.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['location'].fillna(method = 'bfill', limit = 1, inplace = True)


In [118]:
order_z_data

name   
0                                                  Jalsa  \
1                                         Spice Elephant   
2                                        San Churro Cafe   
3                                  Addhuri Udupi Bhojana   
4                                          Grand Village   
...                                                  ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...   
51713                           Vinod Bar And Restaurant   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...   
51716                       The Nest - The Den Bengaluru   

                         location   
0                    Banashankari  \
1                    Banashankari   
2                    Banashankari   
3                    Banashankari   
4                    Basavanagudi   
...                           ...   
51712                  Whitefield   
51713                  Whitefield   
51714                  Whitefield   
51715  ITPL Main Road, Whitefield   
51716  ITPL Main Road, Whitefield   

                                                 address   
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                  phone online_order   
0        080 42297555\r\n+91 9743772233          Yes  \
1                          080 41714161          Yes   
2                        +91 9663487993          Yes   
3                        +91 9620009302           No   
4      +91 8026612447\r\n+91 9901210005           No   
...                                 ...          ...   
51712                      080 40301477           No   
51713                    +91 8197675843           No   
51714                                NM           No   
51715                      080 49652769           No   
51716                    +91 8071117272           No   

                                     cuisines listed_in(type)    rate  votes   
0              North Indian, Mughlai, Chinese          Buffet   4.1/5    775  \
1                 Chinese, North Indian, Thai          Buffet   4.1/5    787   
2                      Cafe, Mexican, Italian          Buffet   3.8/5    918   
3                  South Indian, North Indian          Buffet   3.7/5     88   
4                    North Indian, Rajasthani          Buffet   3.8/5    166   
...                                       ...             ...     ...    ...   
51712                             Continental   Pubs and bars  3.6 /5     27   
51713                             Finger Food   Pubs and bars     NaN      0   
51714                             Finger Food   Pubs and bars     NaN      0   
51715                             Finger Food   Pubs and bars  4.3 /5    236   
51716  Finger Food, North Indian, Continental   Pubs and bars  3.4 /5     13   

      book_table            rest_type approx_cost(for two people)  
0            Yes        Casual Dining                         800  
1             No        Casual Dining                         800  
2             No  Cafe, Casual Dining                         800  
3             No          Quick Bites                         300  
4             No        Casual Dining                         600  
...          ...                  ...                         ...  
51712         No                  Bar                  

In [119]:
# validating by again checking the location column
order_z_data['location'].isnull().sum()

0

**For filling the cuisines column I am taking the same approach to fill as I taken in location column because most of the restaurants are in same location so the chance of same cuisines is more!**

In [125]:
order_z_data['cuisines'].fillna(method = 'bfill', limit = 1, inplace = True)

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/1277404132.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['cuisines'].fillna(method = 'bfill', limit = 1, inplace = True)


In [126]:
order_z_data


name   
0                                                  Jalsa  \
1                                         Spice Elephant   
2                                        San Churro Cafe   
3                                  Addhuri Udupi Bhojana   
4                                          Grand Village   
...                                                  ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...   
51713                           Vinod Bar And Restaurant   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...   
51716                       The Nest - The Den Bengaluru   

                         location   
0                    Banashankari  \
1                    Banashankari   
2                    Banashankari   
3                    Banashankari   
4                    Basavanagudi   
...                           ...   
51712                  Whitefield   
51713                  Whitefield   
51714                  Whitefield   
51715  ITPL Main Road, Whitefield   
51716  ITPL Main Road, Whitefield   

                                                 address   
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                  phone online_order   
0        080 42297555\r\n+91 9743772233          Yes  \
1                          080 41714161          Yes   
2                        +91 9663487993          Yes   
3                        +91 9620009302           No   
4      +91 8026612447\r\n+91 9901210005           No   
...                                 ...          ...   
51712                      080 40301477           No   
51713                    +91 8197675843           No   
51714                                NM           No   
51715                      080 49652769           No   
51716                    +91 8071117272           No   

                                     cuisines listed_in(type)    rate  votes   
0              North Indian, Mughlai, Chinese          Buffet   4.1/5    775  \
1                 Chinese, North Indian, Thai          Buffet   4.1/5    787   
2                      Cafe, Mexican, Italian          Buffet   3.8/5    918   
3                  South Indian, North Indian          Buffet   3.7/5     88   
4                    North Indian, Rajasthani          Buffet   3.8/5    166   
...                                       ...             ...     ...    ...   
51712                             Continental   Pubs and bars  3.6 /5     27   
51713                             Finger Food   Pubs and bars     NaN      0   
51714                             Finger Food   Pubs and bars     NaN      0   
51715                             Finger Food   Pubs and bars  4.3 /5    236   
51716  Finger Food, North Indian, Continental   Pubs and bars  3.4 /5     13   

      book_table            rest_type approx_cost(for two people)  
0            Yes        Casual Dining                         800  
1             No        Casual Dining                         800  
2             No  Cafe, Casual Dining                         800  
3             No          Quick Bites                         300  
4             No        Casual Dining                         600  
...          ...                  ...                         ...  
51712         No                  Bar                  

In [127]:
# validating by checking again the cuisines column 
order_z_data['cuisines'].isnull().sum()

0

In [147]:
order_z_data['rate'] = order_z_data['rate'].str.replace('/5', ' ')

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/2692209636.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['rate'] = order_z_data['rate'].str.replace('/5', ' ')


In [148]:
order_z_data.head()

name      location   
0                  Jalsa  Banashankari  \
1         Spice Elephant  Banashankari   
2        San Churro Cafe  Banashankari   
3  Addhuri Udupi Bhojana  Banashankari   
4          Grand Village  Basavanagudi   

                                             address   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2  1112, Next to KIMS Medical College, 17th Cross...   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   

                              phone online_order   
0    080 42297555\r\n+91 9743772233          Yes  \
1                      080 41714161          Yes   
2                    +91 9663487993          Yes   
3                    +91 9620009302           No   
4  +91 8026612447\r\n+91 9901210005           No   

                         cuisines listed_in(type)  rate  votes book_table   
0  North Indian, Mughlai, Chinese          Buffet  4.1     775        Yes  \
1     Chinese, North Indian, Thai          Buffet  4.1     787         No   
2          Cafe, Mexican, Italian          Buffet  3.8     918         No   
3      South Indian, North Indian          Buffet  3.7      88         No   
4        North Indian, Rajasthani          Buffet  3.8     166         No   

             rest_type approx_cost(for two people)  
0        Casual Dining                         800  
1        Casual Dining                         800  
2  Cafe, Casual Dining                         800  
3          Quick Bites                         300  
4        Casual Dining                         600

### Here I use ffill to fill the missing data point as for now but we can also use the mean and median to do so.
##### To fill the missing values of rate column I am going use the concept of mean.
- Mean or Median Imputation: In this approach, the missing values in the rating column are replaced with the mean or median value of the available ratings. This method assumes that the missing values are missing at random and that the overall distribution of ratings is relatively stable.

In [161]:
order_z_data['rate'].fillna(method = 'ffill', inplace = True)

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/2574308259.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['rate'].fillna(method = 'ffill', inplace = True)


In [163]:
# validating by checking again the rate column
order_z_data['rate'].isnull().sum()

0

In [166]:
order_z_data.dtypes

name                           object
location                       object
address                        object
phone                          object
online_order                   object
cuisines                       object
listed_in(type)                object
rate                           object
votes                           int64
book_table                     object
rest_type                      object
approx_cost(for two people)    object
dtype: object

#### using the same method as for now to fill the missing point of the rest_type column

In [167]:
order_z_data['rest_type'].fillna(method = 'ffill', inplace = True)

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/560888731.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['rest_type'].fillna(method = 'ffill', inplace = True)


In [168]:
# validating by checking the rest_type column
order_z_data['rest_type'].isnull().sum()

0

In [172]:
order_z_data['approx_cost(for two people)'].fillna(method = 'ffill', inplace = True)

C:\Users\kumar\AppData\Local\Temp/ipykernel_19960/2707760037.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  order_z_data['approx_cost(for two people)'].fillna(method = 'ffill', inplace = True)


In [174]:
# validating by checking the  column
order_z_data['approx_cost(for two people)'].isnull().sum()

0

In [175]:
order_z_data

name   
0                                                  Jalsa  \
1                                         Spice Elephant   
2                                        San Churro Cafe   
3                                  Addhuri Udupi Bhojana   
4                                          Grand Village   
...                                                  ...   
51712  Best Brews - Four Points by Sheraton Bengaluru...   
51713                           Vinod Bar And Restaurant   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...   
51716                       The Nest - The Den Bengaluru   

                         location   
0                    Banashankari  \
1                    Banashankari   
2                    Banashankari   
3                    Banashankari   
4                    Basavanagudi   
...                           ...   
51712                  Whitefield   
51713                  Whitefield   
51714                  Whitefield   
51715  ITPL Main Road, Whitefield   
51716  ITPL Main Road, Whitefield   

                                                 address   
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                  phone online_order   
0        080 42297555\r\n+91 9743772233          Yes  \
1                          080 41714161          Yes   
2                        +91 9663487993          Yes   
3                        +91 9620009302           No   
4      +91 8026612447\r\n+91 9901210005           No   
...                                 ...          ...   
51712                      080 40301477           No   
51713                    +91 8197675843           No   
51714                                NM           No   
51715                      080 49652769           No   
51716                    +91 8071117272           No   

                                     cuisines listed_in(type)   rate  votes   
0              North Indian, Mughlai, Chinese          Buffet   4.1     775  \
1                 Chinese, North Indian, Thai          Buffet   4.1     787   
2                      Cafe, Mexican, Italian          Buffet   3.8     918   
3                  South Indian, North Indian          Buffet   3.7      88   
4                    North Indian, Rajasthani          Buffet   3.8     166   
...                                       ...             ...    ...    ...   
51712                             Continental   Pubs and bars  3.6       27   
51713                             Finger Food   Pubs and bars  3.6        0   
51714                             Finger Food   Pubs and bars  3.6        0   
51715                             Finger Food   Pubs and bars  4.3      236   
51716  Finger Food, North Indian, Continental   Pubs and bars  3.4       13   

      book_table            rest_type approx_cost(for two people)  
0            Yes        Casual Dining                         800  
1             No        Casual Dining                         800  
2             No  Cafe, Casual Dining                         800  
3             No          Quick Bites                         300  
4             No        Casual Dining                         600  
...          ...                  ...                         ...  
51712         No                  Bar                       1,500  

In [178]:
order_z_data.isnull().sum()

name                           0
location                       0
address                        0
phone                          0
online_order                   0
cuisines                       0
listed_in(type)                0
rate                           0
votes                          0
book_table                     0
rest_type                      0
approx_cost(for two people)    0
dtype: int64

### Finally I cleaned my dataset my removing the missing data point with valid values

## And the last which I want to do is that I want to change the column name with the meaningfull name

In [181]:
Cleaned_Zomato_Data = order_z_data.set_axis(['Restaurant_Name','Location','Address','Phone_Number','Online_Order','Cuisines','listed_in(type)','Rating','Votes','book_table','Rest_Type','Charge'], axis='columns')

In [182]:
Cleaned_Zomato_Data.head()

Restaurant_Name      Location   
0                  Jalsa  Banashankari  \
1         Spice Elephant  Banashankari   
2        San Churro Cafe  Banashankari   
3  Addhuri Udupi Bhojana  Banashankari   
4          Grand Village  Basavanagudi   

                                             Address   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2  1112, Next to KIMS Medical College, 17th Cross...   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   

                       Phone_Number Online_Order   
0    080 42297555\r\n+91 9743772233          Yes  \
1                      080 41714161          Yes   
2                    +91 9663487993          Yes   
3                    +91 9620009302           No   
4  +91 8026612447\r\n+91 9901210005           No   

                         Cuisines listed_in(type) Rating  Votes book_table   
0  North Indian, Mughlai, Chinese          Buffet   4.1     775        Yes  \
1     Chinese, North Indian, Thai          Buffet   4.1     787         No   
2          Cafe, Mexican, Italian          Buffet   3.8     918         No   
3      South Indian, North Indian          Buffet   3.7      88         No   
4        North Indian, Rajasthani          Buffet   3.8     166         No   

             Rest_Type Charge  
0        Casual Dining    800  
1        Casual Dining    800  
2  Cafe, Casual Dining    800  
3          Quick Bites    300  
4        Casual Dining    600